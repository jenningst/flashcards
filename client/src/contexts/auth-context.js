import React, { createContext, useContext } from 'react';
import Firebase from '../firebase/firebase-client';

const AuthContext = createContext(null);

function AuthProvider(props) {
  const firebase = new Firebase(); // create instance of firebase
  const data = { user: null, }; // initialize our data (for user)
  // provide firebase methods
  const registerWithEmail = (email, password) => firebase.doCreateUserWithEmailAndPassword(email, password);
  const login = (email, password) => firebase.doSignInWithEmailAndPassword(email, password);
  const signOut = () => firebase.doSignOut();
  const resetPasswordWithEmail = (email) => firebase.doResetPassword(email);
  const updatePasswordWithEmail = (password) => firebase.doUpdatePassword(password);

  return (
    <AuthContext.Provider value={
      { data,
        registerWithEmail,
        login,
        signOut,
        resetPasswordWithEmail,
        updatePasswordWithEmail,
      }
    } {...props} />
  );
};

// provides auth context to child components
/* ** IMPLEMENTED IN UserProvider ** */
function useAuth() {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within a AuthContext');
  }
  return context;
};

export { AuthProvider, useAuth };