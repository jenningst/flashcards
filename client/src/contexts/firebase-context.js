import React, { createContext, useContext } from 'react';
import Firebase from '../firebase/firebase';

const FirebaseContext = createContext(null);

// setup our pack provider
function FirebaseProvider({ children }) {
  const firebase = new Firebase();
  return (
    <FirebaseContext.Provider firebase={firebase}>
      {children}
    </FirebaseContext.Provider>
  );
};

// setup our hook to use firebase
function useFirebase() {
  const context = useContext(FirebaseContext);
  if (context === undefined) {
    throw new Error('useFirebase must be used within a FirebaseProvider');
  }
  return context;
};

export { FirebaseProvider, useFirebase };