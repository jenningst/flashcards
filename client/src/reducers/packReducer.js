export const initialState = {
  packFilter: 'SHOW_ALL',
  displayCreatePack: false,
  packMode: '',
};

export default function packReducer(state = initialState, action) {
  switch (action.type) {
    case 'SET_COLLECTION':
      return { ...state, packFilter: action.name };
    case 'TOGGLE_CREATE_PACK':
      return { ...state, displayCreatePack: !state.displayCreatePack };
    case 'CLEAR_COLLECTION':
      return { ...state, packFilter: 'SHOW_ALL' };
    case 'SET_REVIEW_MODE':
      return { ...state, packMode: 'REVIEW_MODE' };
    case 'SET_WRITE_MODE':
      return { ...state, packMode: 'WRITE_MODE' };
    case 'CLEAR_MODE':
      return { ...state, packMode: ''}
    default:
      throw new Error(`Unhandled action type: ${action.type}`);
  };
};
