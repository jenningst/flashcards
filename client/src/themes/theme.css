/* 

button: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      active: '#5860A5',
      disabled: '#e0e0eb',
    },
    font: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    border: {
      primary: '#42487C',
      active: '#939CE8'
      disabled: '#b1aeb7',
      input: '#e9eaf0',
    }, // TODO: consolidate everything below; using the above
    main: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      secondary: '#e3e5ff',
      secondaryHover: '#c3d2ff',
      pureWhite: '#ffffff',
      offWhite: '#f5f7fb',
    },
    alerts: {
      notice: '#c0a5ff',
      success: '#50cdff',
      successHover: '#6FD6FF',
      warning: '#FFCB17',
      warningBg: '#FFDA5C',
      danger: '#DE6A60',
    },
    fonts: {
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    borders: {
      lightGrey: '#d8d6db',
    }

}*/