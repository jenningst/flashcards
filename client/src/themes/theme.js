export const darkTheme = {
  font: {
    primary: '#FFFFFF',
    secondary: '#E0E0E0',
    tertiary: '#666666',
  },
  background: {
    primary: '#282626',
    secondary: '#313131',
    highlight: '#6121BF',
    special: '#7358F7',
    inactive: '#b1aeb7',
  },
  button: {
    default: {
      primary: '#7358F7',
      border: '#4D18AE',
      alert: '#D55949',
      greyed: '#666666',
    },
    disabled: {
      primary: '',
      secondary: '',
      border: '',
    },
    hover: {
      primary: '#6121BF',
      border: '#6121BF',
    },
    active: {
      primary: '#8b90ff',
      border: '',
    },
  },
};

export const lightTheme = {
  color: {
    button: {
      primary: '#939CE8',
      primaryHover: '#7883E2',
      active: '#5860A5',
      secondary: '#50cdff',
      secondaryHover: '#6FD6FF',
      disabled: '#e0e0eb',
    },
    font: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      eerieBlack: '#16171D',
      charleston: '#33364d',
      blackCoral: '#585A6D',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1',
      placeholder: '#e9eaf0',
      danger: '#DE6A60',
    },
    border: {
      primary: '#5860A5',
      secondary: '#7883E2',
      disabled: '#b1aeb7',
      inputDefault: '#e9eaf0',
      transparent: '#0000ffff',
      eerieBlack: '#16171D',
      charleston: '#33364d',
    }, // TODO: consolidate everything below; using the above
    background: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
    },
    main: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      secondary: '#e3e5ff',
      secondaryHover: '#c3d2ff',
      pureWhite: '#ffffff',
      offWhite: '#f5f7fb',
    },
    alerts: {
      notice: '#c0a5ff',
      success: '#50cdff',
      successHover: '#6FD6FF',
      warning: '#FFCB17',
      warningBg: '#FFDA5C',
      danger: '#DE6A60',
    },
    fonts: {
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    borders: {
      lightGrey: '#d8d6db',
    }
  },
  spacing: {
    compact: '1rem',
    comfortable: '1.5rem',
    generous: '2rem',
  },
  font: {
    primary: '#4e5167',
    secondary: '#262626',
    tertiary: '#666666',
  },
  background: {
    primary: '#ffffff',
    secondary: '#fcfbfc',
    highlight: '#9388ff',

    special: '#50cdff',
    inactive: '#b1aeb7',
  },
  button: {
    default: {
      primary: '#7358F7',
      border: '#4D18AE',
      alert: '#D55949',
      greyed: '#666666',
    },
    disabled: {
      primary: '',
      secondary: '',
      border: '',
    },
    hover: {
      primary: '#6121BF',
      border: '#6121BF',
    },
    active: {
      primary: '#8b90ff',
      border: '',
    },
  },
};