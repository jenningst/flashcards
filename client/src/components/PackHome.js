import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { usePackDispatch } from '../contexts/pack-context';
import { ReactComponent as Back } from '../components/icons/svg/back.svg';
import { ReactComponent as EditIcon } from '../components/icons/svg/pencil.svg';
import { ReactComponent as TestIcon } from '../components/icons/svg/play-button.svg';
import { ReactComponent as AddIcon } from '../components/icons/svg/plus.svg';
import { Title2, Caption3 } from './elements/Text';

function PackHome({ name, cards, image }) {
  const dispatch = usePackDispatch();
  const setReviewMode = () => dispatch({ type: 'SET_REVIEW_MODE'});
  const setComposeMode = () => dispatch({ type: 'SET_WRITE_MODE'});

  return (
    <PackHomeWrapper className="PackHome">
      <Header className="PackHome__header">
        <ButtonGroup className="btn-lbl-combo">
          <Link className="btn-lbl-combo__link" to="/">
            <BackIcon
              className="PackHome__button-back btn-lbl-combo__icon"
              data-testid="button-back"
            />
          </Link>
          <Caption3 className="PackHome__button-caption">Back</Caption3>
        </ButtonGroup>
        <CardImage className="PackHome__image-card">
          <img src={image.src} alt={image.alt} />
        </CardImage>
      </Header>
      <PackInfo className="PackHome__info">
        <Title2 className="PackHome__title">
          {name}
        </Title2>
        <Caption3 className="PackHome__count">
          {cards.length === 1 ? `${cards.length} FLASHCARD` : `${cards.length} FLASHCARDS`}
        </Caption3>
      </PackInfo>
      <Footer className="PackHome__footer">
        {
          cards.length > 0
          ? (
              <>
                <ButtonGroup
                  className="btn-lbl-combo"
                  data-testid="test-btn"
                  onClick={setReviewMode}
                >
                  <TestIcon className="PackHome__button-test ico"/>
                  <Caption3>TEST</Caption3>
                </ButtonGroup>

                <ButtonGroup
                  className="btn-lbl-combo"
                  data-testid="edit-btn"
                  onClick={(e) => console.log('clicked edit button')}
                >
                  <EditIcon className="PackHome__button-edit ico" />
                  <Caption3>EDIT</Caption3>
                </ButtonGroup>
              </>
          ):(
            null
          )
        }
        <ButtonGroup
          className="btn-lbl-combo"
          data-testid="add-btn"
          onClick={setComposeMode}
        >
          <AddIcon className="PackHome__button-add ico"/>
          <Caption3>ADD</Caption3>
        </ButtonGroup>
      </Footer>
    </PackHomeWrapper>
  );
};

const PackHomeWrapper = styled.div`
  box-sizing: border-box;
  display: grid;
  grid-template-rows: minmax(0, 1fr) repeat(2, auto);
  grid-row-gap: .50rem;
  height: 100%;
  background: ${props => props.theme.color.main.offWhite};
  color: ${props => props.theme.color.fonts.charleston};
`;

const Header = styled.header`
  display: flex;
  flex-flow: column nowrap;
  padding: 1.5rem;

  div[class~="btn-lbl-combo"] {
    justify-content: flex-start;
    align-items: center;
    margin-bottom: 1rem;

    svg {
      margin-right: .50rem;
    }

    h6 {
      color: ${props => props.theme.color.main.primary};
    }
  }
`;

const ButtonGroup = styled.div`
  display: flex;
  flex-flow: row nowrap;
`;

const BackIcon = styled(Back)`
  height: 2rem;
  width: 2rem;
`;

const CardImage = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100%;
  background: ${props => props.theme.color.main.pureWhite};
  border-radius: 2rem;
  box-shadow: 0px 10px 18px -11px rgba(120,119,120,1);
`;

const PackInfo = styled.section`
  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  align-items: center;
  padding-left: 1.5rem;
  padding-right: 1.5rem;

  .PackHome__title {
    margin-bottom: .25rem;
    font-weight: 500;
    color: ${props => props.theme.color.main.primary};
  }
  .PackHome__count {
    color: ${props => props.theme.color.fonts.grey};
  }
`;

const Footer = styled.footer`
  display: flex;
  justify-content: space-around;
  align-items: center;
  padding: 1.5rem;

  div[class~="btn-lbl-combo"] {
    justify-content: center;
    align-items: center;
    background: ${props => props.theme.color.main.primary};
    color: ${props => props.theme.color.main.pureWhite};
    padding: .75rem 1rem .75rem 1rem;
    border-radius: .50rem;
    box-shadow: 0px 10px 18px -11px rgba(120,119,120,1);

    svg {
      height: 1rem;
      width: 1rem;
      margin-right: .50rem;
      path {
        fill: ${props => props.theme.color.main.pureWhite};
      }
    }
  }
`;

PackHome.propTypes = {
  name: PropTypes.string.isRequired,
  cards: PropTypes.array,
  image: PropTypes.object,
};

PackHome.defaultProps = {
  cards: [],
  image: {
    src: '',
    alt: '',
  },
};

export default PackHome;

