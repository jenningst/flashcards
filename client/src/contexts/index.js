import React from 'react';
import { AuthProvider } from './auth-context';
import { UserProvider } from './user-context';
import { ApolloProvider } from 'react-apollo';

import { ApolloClient, HttpLink } from 'apollo-boost';
import { InMemoryCache } from 'apollo-cache-inmemory';

const client = new ApolloClient({
  link: new HttpLink({ uri: "http://localhost:4000/graphql" }),
  cache: new InMemoryCache(),
});

function AppProviders({ children }) {
  return (
    <AuthProvider>
      <UserProvider>
        <ApolloProvider client={client}>{children}</ApolloProvider>
      </UserProvider>
    </AuthProvider>
  );
};

export default AppProviders;