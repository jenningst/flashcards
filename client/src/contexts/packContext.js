import React, { createContext, useContext, useReducer } from 'react';
import packReducer, { initialState } from '../reducers/packReducer';

const PackStateContext = createContext();
const PackDispatchContext = createContext();

function PackProvider({ children }) {
  const [state, dispatch] = useReducer(packReducer, initialState);
  return (
    <PackStateContext.Provider value={state}>
      <PackDispatchContext.Provider value={dispatch}>
        {children}
      </PackDispatchContext.Provider>
    </PackStateContext.Provider>
  );
};

function usePackState() {
  const context = useContext(PackStateContext);
  if (context === undefined) {
    throw new Error('usePackState must be used within a PackProvider')
  }
  return context;
};

function usePackDispatch() {
  const context = useContext(PackDispatchContext);
  if (context === undefined) {
    throw new Error('usePackDispatch must be used within a PackProvider')
  }
  return context;
};

export { PackProvider, usePackState, usePackDispatch };