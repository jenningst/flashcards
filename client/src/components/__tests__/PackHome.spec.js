import React from 'react';
import { PackProvider } from '../../contexts/packContext';
import { initialState } from '../../reducers/packReducer';
import { MemoryRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import { darkTheme, lightTheme } from '../../themes/theme';

import {
  cleanup,
  render,
} from '@testing-library/react';

import PackHome from '../PackHome';

afterEach(cleanup);

const renderComponent = ({ name, cards, theme }) =>
  render(
    <ThemeProvider theme={theme}>
      <PackProvider value={initialState}>
        <MemoryRouter initialEntries={['/', '/create-pack']}>
          <PackHome name={name} cards={cards} />
        </MemoryRouter>
      </PackProvider>
    </ThemeProvider>
  );

describe('<PackHome /> specs', () => {
  it('assert component matches snapshot', () => {
    const { asFragment } = renderComponent({
      name: 'Random',
      cards: [
        {
          id: '12',
          text: 'What is 2 + 2?',
          answer: '4',
        },
        {
          id: '85',
          text: 'What is the capitol of Sweden?',
          answer: 'Stockholm',
        },
      ],
      theme: lightTheme,
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('assert a pack name is displayed', () => {
    const mockName = 'Random'
    const { getByText } = renderComponent({
      name: mockName,
      cards: [
        {
          id: '12',
          text: 'What is 2 + 2?',
          answer: '4',
        },
        {
          id: '85',
          text: 'What is the capitol of Sweden?',
          answer: 'Stockholm',
        },
      ],
      theme: lightTheme,
    });

    expect(getByText(mockName)).toBeInTheDocument();
  });

  it('assert the correct number of flashcards is displayed', () => {
    const mockCards = [
      {
        id: '12',
        text: 'What is 2 + 2?',
        answer: '4',
      },
      {
        id: '85',
        text: 'What is the capitol of Sweden?',
        answer: 'Stockholm',
      },
    ];
    const { getByText } = renderComponent({
      name: 'Random',
      cards: mockCards,
      theme: lightTheme,
    });
    const numberOfCards = mockCards.length;

    expect(getByText(/flashcard/i)).toHaveTextContent(`${numberOfCards} FLASHCARD`);
  });

  it('assert test button is disabled if no cards in pack', () => {
    const { getByText } = renderComponent({
      name: 'Random',
      cards: [],
      theme: lightTheme,
    });

    expect(getByText(/^test$/i)).toBeDisabled();
  });

  it('assert test button is enabled if 1 or more cards in pack', () => {
    const { getByText } = renderComponent({
      name: 'Random',
      cards: [
        {
          id: '12',
          text: 'What is 2 + 2?',
          answer: '4',
        },
        {
          id: '85',
          text: 'What is the capitol of Sweden?',
          answer: 'Stockholm',
        },
      ],
      theme: lightTheme,
    });

    expect(getByText(/^test$/i)).not.toBeDisabled();
  });

  it('assert compose button is present and enabled', () => {
    const { getByText } = renderComponent({
      name: 'Random',
      cards: [
        {
          id: '12',
          text: 'What is 2 + 2?',
          answer: '4',
        },
        {
          id: '85',
          text: 'What is the capitol of Sweden?',
          answer: 'Stockholm',
        },
      ],
      theme: lightTheme,
    });

    expect(getByText(/^add$/i)).toBeEnabled();
  });

  it.todo('assert clicking `Begin Review` button sends dispatch');
  it.todo('assert clicking `Compose More Cards` button sends dispatch');
});