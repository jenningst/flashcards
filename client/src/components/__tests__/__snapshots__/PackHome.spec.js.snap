// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PackHome /> specs assert component matches snapshot 1`] = `
<div
  class="PackHome sc-jTzLTM iQTdXV"
>
  <header
    class="PackHome__header sc-fjdhpX NXmOK"
  >
    <div
      class="btn-lbl-combo sc-kgoBCf cldxMn"
    >
      <a
        class="btn-lbl-combo__link"
        href="/"
      >
        <svg
          class="PackHome__button-back btn-lbl-combo__icon sc-kGXeez hAVnpz"
          data-testid="button-back"
        >
          back.svg
        </svg>
      </a>
      <h6
        class="PackHome__button-caption sc-VigVT voBwA"
      >
        Back
      </h6>
    </div>
    <svg
      class="sc-kpOJdX kQabda"
      data-testid="edit-btn"
    >
      settings.svg
    </svg>
  </header>
  <section
    class="PackHome__stats stats-section sc-jzJRlG ZMDfA"
  >
    <div
      class="stats-card sc-chPdSV hjTxne"
    >
      <p
        class="stat-section__body sc-dnqmqq eFdqdx"
      >
        Stats Coming Soon!
      </p>
    </div>
  </section>
  <section
    class="PackHome__info sc-cSHVUG bdEAwI"
  >
    <h2
      class="PackHome__title sc-EHOje cayYUN"
    >
      Random
    </h2>
    <div
      class="PackHome__pill sc-dxgOiQ kFfXMj"
    >
      <h6
        class="PackHome__count sc-VigVT voBwA"
      >
        2 FLASHCARDS
      </h6>
    </div>
  </section>
  <section
    class="PackHome__controls sc-kAzzGY hLMvsW"
  >
    <button
      class="sc-bdVaJa btn-lbl-combo sc-bwzfXH ebLKbS"
      data-testid="test-btn"
    >
      TEST
    </button>
    <button
      class="sc-bdVaJa btn-lbl-combo sc-bwzfXH ebLKbS"
      data-testid="add-btn"
    >
      ADD
    </button>
  </section>
</div>
`;
